jetbrains-core/src/software/aws/toolkits/jetbrains/services/cloudwatch/logs/CloudWatchLogStreamCoroutine.kt

```diff
@@ -0,0 +1,133 @@
+            .logStreamName(logStream)
+            .startFromHead(true)
+            .startTime(startTime - duration)
+            .endTime(startTime + duration)
+            .build()
+        val items = load(request, saveForwardToken = true, saveBackwardToken = true)
+        populateTable(items)
+        startListening()
+    }
+
+    private suspend fun populateTable(items: List<OutputLogEvent>) {
+        if (items.isNotEmpty()) {
+            val events = table.logsModel.items.plus(items)
+            withContext(edtContext) { table.logsModel.items = events }
+        }
+        table.emptyText.text = message("cloudwatch.logs.no_events")
+        table.setPaintBusy(false)
+    }
+
+    private suspend fun startListening() = coroutineScope {
+        launch {
+            for (message in channel) {
+                when (message) {
+                    Messages.LOAD_FORWARD -> {
+                        val items = loadMore(nextForwardToken, saveForwardToken = true)
+                        if (items.isNotEmpty()) {
+                            val events = table.logsModel.items.plus(items)
+                            withContext(edtContext) { table.logsModel.items = events }
+                        }
+                    }
+                    Messages.LOAD_BACKWARD -> {
+                        val items = loadMore(nextBackwardToken, saveBackwardToken = true)
+                        if (items.isNotEmpty()) {
+                            val events = items.plus(table.logsModel.items)
+                            withContext(edtContext) { table.logsModel.items = events }
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    private suspend fun loadMore(
+        nextToken: String?,
+        saveForwardToken: Boolean = false,
+        saveBackwardToken: Boolean = false
+    ): List<OutputLogEvent> {
+        val request = GetLogEventsRequest
+            .builder()
+            .logGroupName(logGroup)
+            .logStreamName(logStream)
+            .startFromHead(true)
+            .nextToken(nextToken)
+            .build()
+
+        return load(request, saveForwardToken = saveForwardToken, saveBackwardToken = saveBackwardToken)
+    }
+
+    private suspend fun load(
+        request: GetLogEventsRequest,
+        saveForwardToken: Boolean = false,
+        saveBackwardToken: Boolean = false
+    ): List<OutputLogEvent> {
+        val response = client.getLogEvents(request).asDeferred().await()
+        val events = response.events().filterNotNull()
+        if (saveForwardToken) {
+            nextForwardToken = response.nextForwardToken()
+        }
+        if (saveBackwardToken) {
+            nextBackwardToken = response.nextBackwardToken()
+        }
+
+        return events
+    }
+
+    override fun dispose() {
+        channel.close()
+    }
+
+    private val JBTable.logsModel: ListTableModel<OutputLogEvent> get() = this.model as ListTableModel<OutputLogEvent>
```

Comment:
I would like to avoid this hard cast, can we extend JBTable to simply expose he model better?

Reply:
Did some searching and we can use ` TableView` instead

Reply:
Fixed by changing to `TableView<OutputLogEvent>`